hts221/hts221.go

```diff
@@ -0,0 +1,208 @@
+	err = d.waitForOneShot(0x02)
+	if err != nil {
+		return
+	}
+
+	// read data and calibrate
+	data := []byte{0, 0}
+	d.bus.ReadRegister(d.Address, HTS221_HUMID_OUT_REG, data[:1])
+	d.bus.ReadRegister(d.Address, HTS221_HUMID_OUT_REG+1, data[1:])
+	hValue := readInt(data[1], data[0])
+	hValueCalib := float32(hValue)*d.humiditySlope + d.humidityZero
+
+	return int32(hValueCalib * 1000), nil
+}
+
+// ReadTemperature returns the temperature in celsius milli degrees (°C/1000).
+// Returns an error if the device is not turned on.
+func (d *Device) ReadTemperature() (temperature int32, err error) {
+	err = d.waitForOneShot(0x01)
+	if err != nil {
+		return
+	}
+
+	// read data and calibrate
+	data := []byte{0, 0}
+	d.bus.ReadRegister(d.Address, HTS221_TEMP_OUT_REG, data[:1])
+	d.bus.ReadRegister(d.Address, HTS221_TEMP_OUT_REG+1, data[1:])
+	tValue := readInt(data[1], data[0])
+	tValueCalib := float32(tValue)*d.temperatureSlope + d.temperatureZero
+
+	return int32(tValueCalib * 1000), nil
+}
+
+// Resolution sets the HTS221's resolution mode.
+// The higher resolutions are more accurate but comsume more power (see datasheet).
+// The number of averaged samples will be (h + 2) ^ 2, (t + 1) ^ 2
+//
+func (d *Device) Resolution(h uint8, t uint8) {
+	if h > 7 {
+		h = 3 // default
+	}
+	if t > 7 {
+		t = 3 // default
+	}
+	d.bus.WriteRegister(d.Address, HTS221_AV_CONF_REG, []byte{h<<3 | t})
+}
+
+// private functions
+
+// read factory calibration data
+func (d *Device) calibration() {
+	h0rH, h1rH := []byte{0}, []byte{0}
+	t0degC, t1degC := []byte{0}, []byte{0}
+	t1t0msb := []byte{0}
+	h0t0Out, h1t0Out := []byte{0, 0}, []byte{0, 0}
+	t0Out, t1Out := []byte{0, 0}, []byte{0, 0}
+
+	d.bus.ReadRegister(d.Address, HTS221_H0_rH_x2_REG, h0rH)
+	d.bus.ReadRegister(d.Address, HTS221_H1_rH_x2_REG, h1rH)
+	d.bus.ReadRegister(d.Address, HTS221_T0_degC_x8_REG, t0degC)
+	d.bus.ReadRegister(d.Address, HTS221_T1_degC_x8_REG, t1degC)
+	d.bus.ReadRegister(d.Address, HTS221_T1_T0_MSB_REG, t1t0msb)
+	d.bus.ReadRegister(d.Address, HTS221_H0_T0_OUT_REG, h0t0Out[:1])
+	d.bus.ReadRegister(d.Address, HTS221_H0_T0_OUT_REG+1, h0t0Out[1:])
+	d.bus.ReadRegister(d.Address, HTS221_H1_T0_OUT_REG, h1t0Out[:1])
+	d.bus.ReadRegister(d.Address, HTS221_H1_T0_OUT_REG+1, h1t0Out[1:])
+	d.bus.ReadRegister(d.Address, HTS221_T0_OUT_REG, t0Out[:1])
+	d.bus.ReadRegister(d.Address, HTS221_T0_OUT_REG+1, t0Out[1:])
+	d.bus.ReadRegister(d.Address, HTS221_T1_OUT_REG, t1Out[:1])
+	d.bus.ReadRegister(d.Address, HTS221_T1_OUT_REG+1, t1Out[1:])
+
+	h0rH_v := float32(h0rH[0]) / 2.0
+	h1rH_v := float32(h1rH[0]) / 2.0
+	t0degC_v := float32(readUint(t1t0msb[0]&0x03, t0degC[0])) / 8.0
+	t1degC_v := float32(readUint(t1t0msb[0]&0x0C>>2, t1degC[0])) / 8.0
+	h0t0Out_v := float32(readInt(h0t0Out[1], h0t0Out[0]))
+	h1t0Out_v := float32(readInt(h1t0Out[1], h1t0Out[0]))
+	t0Out_v := float32(readInt(t0Out[1], t0Out[0]))
+	t1Out_v := float32(readInt(t1Out[1], t1Out[0]))
```

Comment:
It is usually better to only use integers, because floating point can be slow. However, in this case it doesn't matter much I think.